enhancements/security/scc-csi-volumes.md

```diff
@@ -0,0 +1,234 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Ephemeral CSI volumes are a newer type of volume in Kubernetes, which allows content to mounted into pods with underlying ephemeral file systems.
+OpenShift Security Context Constraints (SCCs) take an all-or-nothing approach when it comes to these volumes.
+A service account/user can either mount a `csi` volume, or it cannot.
+This proposal will enhance SCCs to control which CSI drivers are allowed.
+
+## Motivation
+
+Inline CSI volumes are provided via CSI drivers that can be added to OpenShift.
+When a pod mounts an inline `csi` volume, it must specify the driver that will provide the underlying content.
+Cluster admins will need a means to identify which drivers are considered safe for use, including in restricted namespaces.
+
+Ephemeral CSI drivers are expected to be used in the following use cases:
+
+1. Mounting shared simple content access certificates used to install RHEL content.
+2. Accessing [sealed secrets](https://secrets-store-csi-driver.sigs.k8s.io/) stored in Vault, Azure Key Vault, or Google Secret Manager.
+
+### Goals
+
+* Allow cluster admins to control which ephemeral CSI drivers can be used to mount `csi` volumes.
+
+### Non-Goals
+
+* Allow `csi` volumes to be mounted in build pods.
+* Enable SCCs to limit [generic ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes)
+
+## Proposal
+
+### User Stories
+
+As an OpenShift cluster admin
+I want to control which CSI drivers can be used to mount csi volumes
+So that my workloads only use trusted CSI drivers to access content
+
+As an application developer
+I want to use the secret store CSI driver to mount csi volumes
+So that I can access sealed secrets in my workloads
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Safe for Consumption Annotation
+
+CSI drivers that provide inline CSI volumes must mark themselves as safe for consumption.
+To mark a CSI driver safe, admins or an operator will need to add the `security.openshift.io/allow-csi-volumes-restricted` annotation to the driver's `CSIDriver` object:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: CSIDriver
+annotations:
+  security.openshift.io/allow-csi-volumes-restricted: "true"
+  ...
+metadata:
+  name: mycsi.dev
+  ...
+```
+
+#### Pod admission
+
+Pod that mount `csi` volumes cannot be admitted unless the chosen SCC allows `csi` volumes.
+In addition, the driver for the volume mount must have the `security.openshift.io/allow-csi-volumes-restricted` annotation.
+
+#### Allow CSI volumes in restricted namespaces
+
+Ephemeral CSI drivers will be promoted to the list of "safe" volume mounts, and be considered equivalent to a Secret or ConfigMap from a safety/restriction perspective.
+CSI volumes will be added to the list of allowed volume types in the `restricted` SCC.
```

Comment:
@deads2k I think we need to be careful here during cluster upgrades. Let's say in 4.9 we allow CSI volumes in the restricted SCC and update openshift-apiserver to check the driver annotation. During upgrade we can have a window where 4.9 SCC manifests are deployed and 4.8 apiservers are still alive. In that situation we would allow inline `csi` volumes on any pod.

What do you think of the following graduation criteria:

1. 4.9 `csi` volumes cannot be mounted unless the driver has the `security.openshift.io/allow-csi-volumes-restricted`  annotation
2. 4.10 we update the default SCCs to allow these volume mounts in restricted namespaces.

Reply:
If we go with an approach that annotates CSI drivers, then no SCC will change.  A 4.8 kube-apiserver will deny the pod.  A 4.9 kube-apiserver would run both admission plugins.

The CSI driver annotation approach means getting the storage team to implement this admission plugin.

Reply:
How does this help us for a pod that uses the restricted SCC? Won't they be denied admission?